MSVS 2019 trips on outputmessage.cpp(82)

<!-- Welcome to the issues section if it's your first time! -->

### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. use MSVS 2019
2. use vcpkg
3. Build master branch

### Expected behaviour
Clean compilation

### Actual behaviour
Compilation error:
```
1>------ Build started: Project: theforgottenserver, Configuration: Debug x64 ------
1>outputmessage.cpp
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664: 'void std::_Normal_allocator_traits<_Alloc>::destroy<_Ty>(_Alloc &,_Ty *)': cannot convert argument 1 from '_Alloc' to '_Alloc &'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:             _Alloc=LockfreePoolingAllocator<OutputMessage,2048>,
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:             _Ty=OutputMessage
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         and
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:             _Alloc=OutputMessageAllocator
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         and
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:             _Alloc=LockfreePoolingAllocator<OutputMessage,2048>
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1580,1): error C2664:         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\xmemory(646,17): message : see declaration of 'std::_Normal_allocator_traits<_Alloc>::destroy'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\xmemory(646,17): message :         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\xmemory(646,17): message :         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\xmemory(646,17): message :             _Alloc=LockfreePoolingAllocator<OutputMessage,2048>
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\xmemory(646,17): message :         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1578): message : while compiling class template member function 'void std::_Ref_count_obj_alloc2<OutputMessage,_Alloc>::_Destroy(void) noexcept'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1578): message :         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1578): message :         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1578): message :             _Alloc=OutputMessageAllocator
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1578): message :         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1609): message : see reference to class template instantiation 'std::_Ref_count_obj_alloc2<OutputMessage,_Alloc>' being compiled
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1609): message :         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1609): message :         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1609): message :             _Alloc=OutputMessageAllocator
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\memory(1609): message :         ]
1>C:\Users\User\code\forgottenserver\src\outputmessage.cpp(82): message : see reference to function template instantiation 'std::shared_ptr<OutputMessage> std::allocate_shared<OutputMessage,OutputMessageAllocator,>(const _Alloc &)' being compiled
1>C:\Users\User\code\forgottenserver\src\outputmessage.cpp(82): message :         with
1>C:\Users\User\code\forgottenserver\src\outputmessage.cpp(82): message :         [
1>C:\Users\User\code\forgottenserver\src\outputmessage.cpp(82): message :             _Alloc=OutputMessageAllocator
1>C:\Users\User\code\forgottenserver\src\outputmessage.cpp(82): message :         ]
1>Done building project "theforgottenserver.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

### Environment
Windows 10 / MSVS 2019

--- a/src/lockfree.h
+++ b/src/lockfree.h
@@ -26,6 +26,13 @@
 
 #include <boost/lockfree/stack.hpp>
 
+/*
+ * we use this to avoid instantiating multiple free lists for objects of the
+ * same size and it can be replaced by a variable template in C++14
+ *
+ * template <std::size_t TSize, size_t CAPACITY>
+ * boost::lockfree::stack<void*, boost::lockfree::capacity<CAPACITY> lockfreeFreeList;
+ */
 template <std::size_t TSize, size_t CAPACITY>
 struct LockfreeFreeList
 {

--- a/src/lockfree.h
+++ b/src/lockfree.h
@@ -30,6 +30,8 @@ template <typename T, size_t CAPACITY>
 class LockfreePoolingAllocator : public std::allocator<T>
 {
 	public:
+		constexpr LockfreePoolingAllocator() {}
+
 		template <typename U>
 		explicit constexpr LockfreePoolingAllocator(const U&) {}
 		using value_type = T;

--- a/src/lockfree.h
+++ b/src/lockfree.h
@@ -26,38 +26,44 @@
 
 #include <boost/lockfree/stack.hpp>
 
+template <std::size_t TSize, size_t CAPACITY>
+struct LockfreeFreeList
+{
+	using FreeList = boost::lockfree::stack<void*, boost::lockfree::capacity<CAPACITY>>;
+	static FreeList& get()
+	{
+		static FreeList freeList;
+		return freeList;
+	}
+};
+
 template <typename T, size_t CAPACITY>
 class LockfreePoolingAllocator : public std::allocator<T>
 {
 	public:
-		constexpr LockfreePoolingAllocator() {}
+		LockfreePoolingAllocator() = default;
 
-		template <typename U>
+		template <typename U, class = typename std::enable_if<!std::is_same<U, T>::value>::type>
 		explicit constexpr LockfreePoolingAllocator(const U&) {}
 		using value_type = T;
 
 		T* allocate(size_t) const {
-			T* p; // NOTE: p doesn't have to be initialized
-			if (!getFreeList().pop(p)) {
+			auto& inst = LockfreeFreeList<sizeof(T), CAPACITY>::get();
+			void* p; // NOTE: p doesn't have to be initialized
+			if (!inst.pop(p)) {
 				//Acquire memory without calling the constructor of T
-				p = static_cast<T*>(operator new (sizeof(T)));
+				p = operator new (sizeof(T));
 			}
-			return p;
+			return static_cast<T*>(p);
 		}
 
 		void deallocate(T* p, size_t) const {
-			if (!getFreeList().bounded_push(p)) {
+			auto& inst = LockfreeFreeList<sizeof(T), CAPACITY>::get();
+			if (!inst.bounded_push(p)) {
 				//Release memory without calling the destructor of T
 				//(it has already been called at this point)
 				operator delete(p);
 			}
-		}
-
-	private:
-		using FreeList = boost::lockfree::stack<T*, boost::lockfree::capacity<CAPACITY>>;
-		static FreeList& getFreeList() {
-			static FreeList freeList;
-			return freeList;
 		}
 };
 

--- a/src/outputmessage.cpp
+++ b/src/outputmessage.cpp
@@ -29,15 +29,6 @@ extern Scheduler g_scheduler;
 const uint16_t OUTPUTMESSAGE_FREE_LIST_CAPACITY = 2048;
 const std::chrono::milliseconds OUTPUTMESSAGE_AUTOSEND_DELAY {10};
 
-template<typename U>
-using InternalOutputMessageAllocator = LockfreePoolingAllocator<U, OUTPUTMESSAGE_FREE_LIST_CAPACITY>;
-class OutputMessageAllocator : public InternalOutputMessageAllocator<OutputMessage>
-{
-	public:
-		template<typename U>
-		struct rebind {using other = InternalOutputMessageAllocator<U>;};
-};
-
 void OutputMessagePool::scheduleSendAll()
 {
 	auto functor = std::bind(&OutputMessagePool::sendAll, this);
@@ -80,5 +71,7 @@ void OutputMessagePool::removeProtocolFromAutosend(const Protocol_ptr& protocol)
 
 OutputMessage_ptr OutputMessagePool::getOutputMessage()
 {
-	return std::allocate_shared<OutputMessage>(OutputMessageAllocator());
+	// LockfreePoolingAllocator<void,...> will leave (void* allocate) ill-formed because
+	// of sizeof(T), so this guaranatees that only one list will be initialized
+	return std::allocate_shared<OutputMessage>(LockfreePoolingAllocator<void, OUTPUTMESSAGE_FREE_LIST_CAPACITY>());
 }

--- a/src/outputmessage.cpp
+++ b/src/outputmessage.cpp
@@ -29,12 +29,13 @@ extern Scheduler g_scheduler;
 const uint16_t OUTPUTMESSAGE_FREE_LIST_CAPACITY = 2048;
 const std::chrono::milliseconds OUTPUTMESSAGE_AUTOSEND_DELAY {10};
 
-class OutputMessageAllocator
+template<typename U>
+using InternalOutputMessageAllocator = LockfreePoolingAllocator<U, OUTPUTMESSAGE_FREE_LIST_CAPACITY>;
+class OutputMessageAllocator : public InternalOutputMessageAllocator<OutputMessage>
 {
 	public:
-		using value_type = OutputMessage;
 		template<typename U>
-		struct rebind {using other = LockfreePoolingAllocator<U, OUTPUTMESSAGE_FREE_LIST_CAPACITY>;};
+		struct rebind {using other = InternalOutputMessageAllocator<U>;};
 };
 
 void OutputMessagePool::scheduleSendAll()

